File: autocontext/src/autocontext/runtimes/pi_rpc.py
"""Pi RPC runtime — stdin/stdout JSONL subprocess communication (AC-375).

Pi RPC mode (`pi --mode rpc`) communicates over the process's stdin/stdout
using strict JSONL framing (LF-delimited). This is NOT an HTTP protocol.

Protocol reference:
  https://github.com/badlogic/pi-mono/blob/main/packages/coding-agent/docs/rpc.md
"""

from __future__ import annotations

import json
import logging
import queue
import shutil
import subprocess
import threading
import time
import uuid
from dataclasses import dataclass, field
from typing import Any

from autocontext.runtimes.base import AgentOutput, AgentRuntime

logger = logging.getLogger(__name__)


@dataclass(slots=True)
class PiRPCConfig:
    """Configuration for the Pi RPC runtime.

    Pi RPC is a subprocess protocol over stdin/stdout JSONL — not HTTP.
    The ``endpoint`` field is intentionally absent.
    """

    pi_command: str = "pi"
    model: str = ""
    timeout: float = 120.0
    session_persistence: bool = True
    no_context_files: bool = False
    branch_on_retry: bool = True
    extra_args: list[str] = field(default_factory=list)


class PiRPCRuntime(AgentRuntime):
    """Agent runtime that communicates with Pi via stdin/stdout RPC.

    Launches ``pi --mode rpc`` as a subprocess and exchanges JSONL
    messages on stdin/stdout per Pi's documented RPC protocol.
    """

    def __init__(self, config: PiRPCConfig | None = None) -> None:
        self._config = config or PiRPCConfig()
        self._pi_path = shutil.which(self._config.pi_command)
        self._current_session_id: str | None = None

    @property
    def available(self) -> bool:
        return self._pi_path is not None

    def _build_args(self) -> list[str]:
        """Build the pi --mode rpc argument list."""
        pi = self._pi_path or self._config.pi_command
        args = [pi, "--mode", "rpc"]
        if self._config.model:
            args.extend(["--model", self._config.model])
        if self._config.no_context_files:
            args.append("--no-context-files")
        if not self._config.session_persistence:
            args.append("--no-session")
        args.extend(self._config.extra_args)
        return args

    def _build_prompt_command(self, prompt: str) -> dict[str, Any]:
        """Build a Pi RPC prompt command.

        Pi's documented RPC protocol expects the user payload under ``message``.
        """
        return {
            "type": "prompt",
            "id": uuid.uuid4().hex[:8],
            "message": prompt,
        }

    def _nonzero_exit_output(self, exit_code: int, stderr: str, stdout: str = "") -> AgentOutput:
        metadata: dict[str, Any] = {
            "error": "nonzero_exit",
            "exit_code": exit_code,
        }
        if stderr:
            metadata["stderr"] = stderr[:500]
        if stdout:
            metadata["stdout"] = stdout[:500]
        return AgentOutput(text="", metadata=metadata)

    def _read_stream(self, stream: Any, sink: queue.Queue[str | None]) -> None:
        try:
            for line in stream:
                sink.put(line)
        finally:
            sink.put(None)

    def _drain_queue(self, source: queue.Queue[str | None], lines: list[str]) -> None:
        while True:
            try:
                line = source.get_nowait()
            except queue.Empty:
                return
            if line is not None:
                lines.append(line)

    def _is_terminal_rpc_event(self, line: str) -> bool:
        try:
            event = json.loads(line)
        except (json.JSONDecodeError, TypeError):
            return False
        if not isinstance(event, dict):
            return False
        if event.get("type") == "agent_end":
            return True
        return event.get("type") == "response" and event.get("success") is False

    def _shutdown_process(self, process: subprocess.Popen[str]) -> None:
        if process.stdin and not process.stdin.closed:
            try:
                process.stdin.close()
            except OSError:
                logger.debug("runtimes.pi_rpc: suppressed stdin close error", exc_info=True)
        try:
            process.wait(timeout=1.0)
        except subprocess.TimeoutExpired:
            process.terminate()
            try:
                process.wait(timeout=1.0)
            except subprocess.TimeoutExpired:
                process.kill()
                process.wait(timeout=1.0)

    def generate(
        self,
        prompt: str,
        system: str | None = None,
        schema: dict | None = None,
    ) -> AgentOutput:
        """Send a prompt command and collect the response."""
        full_prompt = prompt
        if system:
            full_prompt = f"{system}\n\n{prompt}"

        command = self._build_prompt_command(full_prompt)

        args = self._build_args()
        stdout_queue: queue.Queue[str | None] = queue.Queue()
        stderr_queue: queue.Queue[str | None] = queue.Queue()
        stdout_lines: list[str] = []
        stderr_lines: list[str] = []
        process: subprocess.Popen[str] | None = None
        try:
            # Keep stdin open after the prompt ack so Pi can stream the agent result.
            input_line = json.dumps(command) + "\n"
            process = subprocess.Popen(
                args,
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
            )
            if process.stdin is None or process.stdout is None or process.stderr is None:
                return AgentOutput(text="", metadata={"error": "pi_rpc_pipe_unavailable"})

            threading.Thread(target=self._read_stream, args=(process.stdout, stdout_queue), daemon=True).start()
            threading.Thread(target=self._read_stream, args=(process.stderr, stderr_queue), daemon=True).start()

            process.stdin.write(input_line)
            process.stdin.flush()

            deadline = time.monotonic() + self._config.timeout
            saw_stdout_eof = False
            terminal_seen = False
            while True:
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    raise subprocess.TimeoutExpired(args, self._config.timeout)

                try:
                    line = stdout_queue.get(timeout=min(0.1, remaining))
                except queue.Empty:
                    self._drain_queue(stderr_queue, stderr_lines)
                    if process.poll() is not None and saw_stdout_eof:
                        break
                    continue

                if line is None:
                    saw_stdout_eof = True
                    if process.poll() is not None:
                        break
                    continue

                stdout_lines.append(line)
                if self._is_terminal_rpc_event(line):
                    terminal_seen = True
                    break

            self._drain_queue(stderr_queue, stderr_lines)
            if terminal_seen:
                self._shutdown_process(process)
            else:
                process.wait(timeout=1.0)
        except subprocess.TimeoutExpired:
            logger.error("pi RPC timed out after %.0fs", self._config.timeout)
            if process is not None:
                process.kill()
                process.wait(timeout=1.0)
            return AgentOutput(text="", metadata={"error": "timeout"})
        except FileNotFoundError:
            logger.error("pi CLI not found at %r", self._config.pi_command)
            return AgentOutput(text="", metadata={"error": "pi_not_found"})

        stdout = "".join(stdout_lines)
        stderr = "".join(stderr_lines)
        returncode = process.returncode if process is not None and process.returncode is not None else 0
        if returncode != 0 and not stdout.strip():
            logger.warning("pi RPC exited with code %d: %s", returncode, stderr[:200])
            return self._nonzero_exit_output(returncode, stderr)

        output = self._parse_rpc_output(
            stdout,
            exit_code=returncode,
            stderr=stderr,
        )
        if returncode != 0 and not output.metadata.get("error"):
            logger.warning("pi RPC exited with code %d: %s", returncode, stderr[:200])
            return self._nonzero_exit_output(returncode, stderr, stdout)
        return output

    def revise(
        self,
        prompt: str,
        previous_output: str,
        feedback: str,
        system: str | None = None,
    ) -> AgentOutput:
        revision_prompt = (
            f"Revise the following output based on the judge's feedback.\n\n"
            f"## Original Output\n{previous_output}\n\n"
            f"## Judge Feedback\n{feedback}\n\n"
            f"## Original Task\n{prompt}\n\n"
            "Produce an improved version:"
        )
        return self.generate(revision_prompt, system=system)

    def _parse_rpc_output(
        self,
        raw: str,
        *,
        exit_code: int = 0,
        stderr: str = "",
    ) -> AgentOutput:
        """Parse JSONL output from pi --mode rpc.

        Collects message_end events to extract the assistant's response.
        Falls back to the last non-empty line if no structured events found.
        """
        text_parts: list[str] = []
        saw_json_event = False

        for line in raw.strip().split("\n"):
            line = line.strip()
            if not line:
                continue
            try:
                event = json.loads(line)
                saw_json_event = True
                event_type = event.get("type", "")

                # Collect assistant text from message_end or response events
                if event_type == "response":
                    if event.get("success") is False:
                        metadata: dict[str, Any] = {
                            "error": "rpc_response_error",
                            "rpc_command": str(event.get("command", "")),
                            "exit_code": exit_code,
                        }
                        error_message = event.get("error")
                        if error_message is not None:
                            metadata["rpc_message"] = str(error_message)
                        if stderr:
                            metadata["stderr"] = stderr[:500]
                        return AgentOutput(text="", metadata=metadata)

                    data = event.get("data", {})
                    if isinstance(data, dict) and "content" in data:
                        text_parts.append(str(data["content"]))
                elif event_type == "message_end":
                    msg = event.get("message", {})
                    content = msg.get("content", "")
                    if isinstance(content, str) and content:
                        text_parts.append(content)
                elif event_type == "agent_end":
                    # Final messages array
                    messages = event.get("messages", [])
                    for msg in messages:
                        if isinstance(msg, dict) and msg.get("role") == "assistant":
                            content = msg.get("content", "")
                            if isinstance(content, str) and content:
                                text_parts.append(content)
            except (json.JSONDecodeError, TypeError):
                # Not JSONL — treat as plain text fallback
                if not text_parts:
                    if exit_code != 0:
                        return self._nonzero_exit_output(exit_code, stderr, raw.strip())
                    return AgentOutput(text=raw.strip(), metadata={"exit_code": exit_code})

        if text_parts:
            return AgentOutput(text=text_parts[-1], metadata={"exit_code": exit_code})  # Last assistant message

        if exit_code != 0:
            return self._nonzero_exit_output(exit_code, stderr, raw.strip())
        if saw_json_event:
            return AgentOutput(
                text="",
                metadata={
                    "error": "missing_assistant_response",
                    "exit_code": exit_code,
                    "stdout": raw.strip()[:500],
                },
            )
        return AgentOutput(text=raw.strip(), metadata={"exit_code": exit_code})

File: ts/src/runtimes/pi-rpc.ts
/**
 * Pi RPC runtime — subprocess stdin/stdout JSONL communication with Pi.
 * Mirrors Python's autocontext/runtimes/pi_rpc.py.
 */

import { randomUUID } from "node:crypto";
import { spawn } from "node:child_process";
import type { AgentOutput } from "./base.js";
import { definedConfigOptions } from "./config-options.js";

export interface PiRPCConfigOpts {
  piCommand?: string;
  model?: string;
  timeout?: number;
  workspace?: string;
  sessionPersistence?: boolean;
  noContextFiles?: boolean;
  extraArgs?: string[];
}

const PI_RPC_CONFIG_DEFAULTS = {
  piCommand: "pi",
  model: "",
  timeout: 120.0,
  workspace: "",
  sessionPersistence: true,
  noContextFiles: false,
  extraArgs: [] as string[],
};

export class PiRPCConfig {
  readonly piCommand!: string;
  readonly model!: string;
  readonly timeout!: number;
  readonly workspace!: string;
  readonly sessionPersistence!: boolean;
  readonly noContextFiles!: boolean;
  readonly extraArgs!: string[];

  constructor(opts: PiRPCConfigOpts = {}) {
    Object.assign(this, {
      ...PI_RPC_CONFIG_DEFAULTS,
      ...definedConfigOptions(opts),
      extraArgs: [...(opts.extraArgs ?? PI_RPC_CONFIG_DEFAULTS.extraArgs)],
    });
  }
}

export class PiRPCRuntime {
  readonly name = "pi-rpc";
  protected config: PiRPCConfig;
  protected _currentSessionId: string | null = null;

  constructor(config?: PiRPCConfig) {
    this.config = config ?? new PiRPCConfig();
  }

  get currentSessionId(): string | null {
    return this._currentSessionId;
  }

  async generate(opts: { prompt: string; system?: string }): Promise<AgentOutput> {
    const fullPrompt = opts.system ? `${opts.system}\n\n${opts.prompt}` : opts.prompt;
    const args = this.buildArgs();
    const input = `${JSON.stringify(this.buildPromptCommand(fullPrompt))}\n`;

    return this.invokeRpc(args, input);
  }

  async revise(opts: {
    prompt: string;
    previousOutput: string;
    feedback: string;
  }): Promise<AgentOutput> {
    return this.generate({
      prompt: [
        `Revise the following output based on the judge's feedback.`,
        `## Original Output\n${opts.previousOutput}`,
        `## Judge Feedback\n${opts.feedback}`,
        `## Original Task\n${opts.prompt}`,
        `Produce an improved version:`,
      ].join("\n\n"),
    });
  }

  protected buildArgs(): string[] {
    const args = ["--mode", "rpc"];
    if (this.config.model) {
      args.push("--model", this.config.model);
    }
    if (this.config.noContextFiles) {
      args.push("--no-context-files");
    }
    if (!this.config.sessionPersistence) {
      args.push("--no-session");
    }
    args.push(...this.config.extraArgs);
    return args;
  }

  protected buildPromptCommand(prompt: string): { type: string; id: string; message: string } {
    return {
      type: "prompt",
      id: randomUUID().slice(0, 8),
      message: prompt,
    };
  }

  private invokeRpc(args: string[], input: string): Promise<AgentOutput> {
    return new Promise((resolve) => {
      const child = spawn(this.config.piCommand, args, {
        stdio: ["pipe", "pipe", "pipe"],
        cwd: this.config.workspace || undefined,
      });
      let stdout = "";
      let stderr = "";
      let stdoutBuffer = "";
      let settled = false;

      const cleanupTimer = (): NodeJS.Timeout =>
        setTimeout(() => {
          if (!child.killed && child.exitCode === null) {
            child.kill();
          }
        }, 1_000);

      const finish = (output: AgentOutput, endStdin = true): void => {
        if (settled) return;
        settled = true;
        clearTimeout(timeout);
        if (endStdin && child.stdin.writable && !child.stdin.destroyed) {
          child.stdin.end();
        }
        cleanupTimer().unref();
        resolve(output);
      };

      const timeout = setTimeout(() => {
        if (!child.killed) {
          child.kill();
        }
        finish({ text: "", metadata: { error: "timeout" } }, false);
      }, this.config.timeout * 1000);

      child.stdout.setEncoding("utf-8");
      child.stderr.setEncoding("utf-8");

      child.stdout.on("data", (chunk: string | Buffer) => {
        stdoutBuffer += this.normalizeOutput(chunk);
        let newlineIndex = stdoutBuffer.indexOf("\n");
        while (newlineIndex >= 0) {
          const line = stdoutBuffer.slice(0, newlineIndex);
          stdoutBuffer = stdoutBuffer.slice(newlineIndex + 1);
          stdout += `${line}\n`;
          if (this.isTerminalRpcEvent(line)) {
            finish(this.parseOutput(stdout, 0, stderr));
            return;
          }
          newlineIndex = stdoutBuffer.indexOf("\n");
        }
      });

      child.stderr.on("data", (chunk: string | Buffer) => {
        stderr += this.normalizeOutput(chunk);
      });

      child.on("error", (err: NodeJS.ErrnoException) => {
        if (err.code === "ENOENT") {
          finish({ text: "", metadata: { error: "pi_not_found" } }, false);
          return;
        }
        finish({ text: "", metadata: { error: err.message || "unknown" } }, false);
      });

      child.on("close", (code) => {
        if (stdoutBuffer) {
          stdout += stdoutBuffer;
          stdoutBuffer = "";
        }
        finish(this.parseOutput(stdout, code ?? 1, stderr), false);
      });

      child.stdin.write(input);
    });
  }

  protected isTerminalRpcEvent(record: string): boolean {
    try {
      const event = JSON.parse(record) as {
        type?: string;
        success?: boolean;
      };
      if (event.type === "agent_end") return true;
      return event.type === "response" && event.success === false;
    } catch {
      return false;
    }
  }

  protected parseOutput(raw: string, exitCode: number, stderr: string): AgentOutput {
    const trimmed = raw.trim();
    if (!trimmed) {
      return exitCode === 0
        ? { text: "", metadata: { exitCode } }
        : {
            text: "",
            metadata: {
              error: "nonzero_exit",
              exitCode,
              stderr,
            },
          };
    }

    const textParts: string[] = [];
    let sawJsonEvent = false;

    for (const line of trimmed.split("\n")) {
      const record = line.trim();
      if (!record) continue;

      try {
        const event = JSON.parse(record) as {
          type?: string;
          success?: boolean;
          command?: string;
          error?: unknown;
          data?: { content?: unknown; session_id?: unknown; sessionId?: unknown };
          message?: { content?: unknown };
          messages?: Array<{ role?: string; content?: unknown }>;
          session_id?: unknown;
          sessionId?: unknown;
        };
        sawJsonEvent = true;
        this.updateSessionId(event);

        if (event.type === "response") {
          if (event.success === false) {
            return {
              text: "",
              metadata: {
                error: "rpc_response_error",
                rpcCommand: String(event.command ?? ""),
                rpcMessage: String(event.error ?? "unknown"),
                exitCode,
                ...(stderr ? { stderr } : {}),
              },
            };
          }

          const content = this.extractTextContent(event.data?.content);
          if (content) {
            textParts.push(content);
          }
          continue;
        }

        if (event.type === "message_end") {
          const content = this.extractTextContent(event.message?.content);
          if (content) {
            textParts.push(content);
          }
          continue;
        }

        if (event.type === "agent_end") {
          for (const message of event.messages ?? []) {
            if (message.role === "assistant") {
              const content = this.extractTextContent(message.content);
              if (content) {
                textParts.push(content);
              }
            }
          }
        }
      } catch {
        if (textParts.length === 0) {
          return exitCode === 0
            ? { text: trimmed, metadata: { exitCode } }
            : {
                text: "",
                metadata: {
                  error: "nonzero_exit",
                  exitCode,
                  ...(stderr ? { stderr } : {}),
                  stdout: trimmed,
                },
              };
        }
      }
    }

    if (textParts.length > 0) {
      return {
        text: textParts[textParts.length - 1] ?? "",
        metadata: {
          exitCode,
          ...(this._currentSessionId ? { sessionId: this._currentSessionId } : {}),
        },
      };
    }

    return exitCode === 0
      ? sawJsonEvent
        ? {
            text: "",
            metadata: {
              error: "missing_assistant_response",
              exitCode,
              stdout: trimmed,
            },
          }
        : { text: trimmed, metadata: { exitCode } }
      : {
          text: "",
          metadata: {
            error: "nonzero_exit",
            exitCode,
            ...(stderr ? { stderr } : {}),
            stdout: trimmed,
          },
        };
  }

  protected extractTextContent(content: unknown): string {
    if (typeof content === "string") {
      return content;
    }
    if (!Array.isArray(content)) {
      return "";
    }
    return content
      .map((part) => {
        if (typeof part === "string") return part;
        if (!part || typeof part !== "object") return "";
        if ("text" in part && typeof part.text === "string") return part.text;
        if ("content" in part && typeof part.content === "string") return part.content;
        return "";
      })
      .filter(Boolean)
      .join("");
  }

  protected updateSessionId(event: {
    data?: { session_id?: unknown; sessionId?: unknown };
    session_id?: unknown;
    sessionId?: unknown;
  }): void {
    const candidate =
      event.data?.session_id ?? event.data?.sessionId ?? event.session_id ?? event.sessionId;
    if (typeof candidate === "string" && candidate) {
      this._currentSessionId = candidate;
    }
  }

  protected normalizeOutput(value: string | Buffer | undefined): string {
    if (typeof value === "string") {
      return value;
    }
    if (value) {
      return value.toString("utf-8");
    }
    return "";
  }
}

type PiRpcCommand = Record<string, unknown> & { type: string; id?: string };
type PiRpcEvent = Record<string, unknown> & { type?: string };

export class PiPersistentRPCRuntime extends PiRPCRuntime {
  readonly supportsConcurrentRequests = false;
  #process: ReturnType<typeof spawn> | null = null;
  #stdoutBuffer = "";
  #stdoutLines: string[] = [];
  #stderr = "";
  #waiters: Array<() => void> = [];
  #processError: Error | null = null;
  #processExitCode: number | null = null;

  close(): void {
    const child = this.#process;
    if (!child) return;
    if (child.stdin && child.stdin.writable && !child.stdin.destroyed) {
      child.stdin.end();
    }
    if (!child.killed && child.exitCode === null) {
      child.kill();
    }
    this.#process = null;
    this.notifyWaiters();
  }

  override async generate(opts: {
    prompt: string;
    system?: string;
    schema?: Record<string, unknown>;
  }): Promise<AgentOutput> {
    void opts.schema;
    const fullPrompt = opts.system ? `${opts.system}\n\n${opts.prompt}` : opts.prompt;
    const command = this.withId(this.buildPromptCommand(fullPrompt));
    try {
      const lines = await this.collectUntil(command, (line) => this.isTerminalRpcEvent(line));
      return this.parseOutput(lines.join(""), this.#processExitCode ?? 0, this.#stderr);
    } catch (error) {
      if (error instanceof Error && error.message === "timeout") {
        this.close();
        return { text: "", metadata: { error: "timeout" } };
      }
      this.close();
      return { text: "", metadata: { error: error instanceof Error ? error.message : String(error) } };
    }
  }

  async steer(message: string): Promise<Record<string, unknown>> {
    return this.collectResponse({ type: "steer", message });
  }

  async followUp(message: string): Promise<Record<string, unknown>> {
    return this.collectResponse({ type: "follow_up", message });
  }

  async abort(): Promise<Record<string, unknown>> {
    return this.collectResponse({ type: "abort" });
  }

  async getState(): Promise<Record<string, unknown>> {
    const response = await this.collectResponse({ type: "get_state" });
    const { success: _success, ...state } = response;
    return state;
  }

  async getMessages(): Promise<Array<Record<string, unknown>>> {
    const response = await this.collectResponse({ type: "get_messages" });
    return Array.isArray(response.messages)
      ? response.messages.filter((message): message is Record<string, unknown> =>
          Boolean(message) && typeof message === "object" && !Array.isArray(message),
        )
      : [];
  }

  private ensureProcess(): ReturnType<typeof spawn> {
    if (this.#process && this.#process.exitCode === null && !this.#process.killed) {
      return this.#process;
    }
    this.#stdoutBuffer = "";
    this.#stdoutLines = [];
    this.#stderr = "";
    this.#processError = null;
    this.#processExitCode = null;

    const child = spawn(this.config.piCommand, this.buildArgs(), {
      stdio: ["pipe", "pipe", "pipe"],
      cwd: this.config.workspace || undefined,
    });
    child.stdout.setEncoding("utf-8");
    child.stderr.setEncoding("utf-8");
    child.stdout.on("data", (chunk: string | Buffer) => {
      this.pushStdout(chunk);
    });
    child.stderr.on("data", (chunk: string | Buffer) => {
      this.#stderr += this.normalizeOutput(chunk);
    });
    child.on("error", (error) => {
      this.#processError = error instanceof Error ? error : new Error(String(error));
      this.notifyWaiters();
    });
    child.on("close", (code, signal) => {
      if (this.#stdoutBuffer) {
        this.#stdoutLines.push(this.#stdoutBuffer);
        this.#stdoutBuffer = "";
      }
      this.#processExitCode = code ?? (signal ? 1 : 0);
      this.notifyWaiters();
    });
    this.#process = child;
    return child;
  }

  private pushStdout(chunk: string | Buffer): void {
    this.#stdoutBuffer += this.normalizeOutput(chunk);
    let newlineIndex = this.#stdoutBuffer.indexOf("\n");
    while (newlineIndex >= 0) {
      const line = this.#stdoutBuffer.slice(0, newlineIndex);
      this.#stdoutBuffer = this.#stdoutBuffer.slice(newlineIndex + 1);
      this.#stdoutLines.push(`${line}\n`);
      newlineIndex = this.#stdoutBuffer.indexOf("\n");
    }
    this.notifyWaiters();
  }

  private writeCommand(command: PiRpcCommand): void {
    const child = this.ensureProcess();
    const stdin = child.stdin;
    if (!stdin || !stdin.writable || stdin.destroyed) {
      throw new Error("pi RPC stdin unavailable");
    }
    stdin.write(`${JSON.stringify(command)}\n`);
  }

  private async collectUntil(command: PiRpcCommand, terminal: (line: string) => boolean): Promise<string[]> {
    this.writeCommand(command);
    const deadline = Date.now() + this.config.timeout * 1000;
    const lines: string[] = [];
    while (true) {
      const line = await this.nextLine(deadline);
      if (line === null) break;
      lines.push(line);
      if (terminal(line)) break;
    }
    return lines;
  }

  private async collectResponse(command: PiRpcCommand): Promise<Record<string, unknown>> {
    const resolved = this.withId(command);
    const lines = await this.collectUntil(resolved, (line) => {
      const event = this.loadEvent(line);
      return Boolean(event && this.isResponseFor(event, resolved));
    });

    for (const line of [...lines].reverse()) {
      const event = this.loadEvent(line);
      if (!event || !this.isResponseFor(event, resolved)) continue;
      if (event.success === false) {
        return {
          success: false,
          error: event.error ?? "",
          command: event.command ?? resolved.type,
        };
      }
      const data = event.data;
      return this.isRecord(data)
        ? { success: true, ...data }
        : { success: true, data };
    }

    return { success: false, error: "missing_rpc_response", command: resolved.type };
  }

  private async nextLine(deadline: number): Promise<string | null> {
    if (this.#stdoutLines.length > 0) {
      return this.#stdoutLines.shift() ?? null;
    }
    if (this.#processError) {
      throw this.#processError;
    }
    if (this.#processExitCode !== null) {
      return null;
    }

    const remaining = deadline - Date.now();
    if (remaining <= 0) {
      throw new Error("timeout");
    }
    await this.waitForOutput(remaining);
    if (this.#stdoutLines.length > 0) {
      return this.#stdoutLines.shift() ?? null;
    }
    if (this.#processError) {
      throw this.#processError;
    }
    if (Date.now() >= deadline) {
      throw new Error("timeout");
    }
    return null;
  }

  private waitForOutput(timeoutMs: number): Promise<void> {
    return new Promise((resolve) => {
      let settled = false;
      const timer = setTimeout(() => {
        if (settled) return;
        settled = true;
        this.#waiters = this.#waiters.filter((waiter) => waiter !== notify);
        resolve();
      }, timeoutMs);
      const notify = (): void => {
        if (settled) return;
        settled = true;
        clearTimeout(timer);
        resolve();
      };
      this.#waiters.push(notify);
    });
  }

  private notifyWaiters(): void {
    const waiters = this.#waiters.splice(0);
    for (const waiter of waiters) {
      waiter();
    }
  }

  private withId<T extends PiRpcCommand>(command: T): T {
    return command.id ? command : { ...command, id: randomUUID().slice(0, 8) };
  }

  private loadEvent(line: string): PiRpcEvent | null {
    try {
      const event: unknown = JSON.parse(line);
      return this.isRecord(event) ? event : null;
    } catch {
      return null;
    }
  }

  private isRecord(value: unknown): value is PiRpcEvent {
    return Boolean(value) && typeof value === "object" && !Array.isArray(value);
  }

  private isResponseFor(event: PiRpcEvent, command: PiRpcCommand): boolean {
    if (event.type !== "response") return false;
    if (typeof event.id === "string" && command.id) {
      return event.id === command.id;
    }
    return event.command === command.type;
  }
}
